subsys/net/conn_mgr/conn_mgr_monitor.c

```diff
@@ -91,6 +143,537 @@ static void conn_mgr_mon_set_ready(int idx, bool readiness)
+		}
+	}
+
+	return 0;
+}
+
+static void response_cb(struct http_response *rsp,
+			enum http_final_call final_data,
+			void *user_data)
+{
+	uint16_t *status = user_data;
+
+	if (final_data == HTTP_DATA_MORE) {
+		NET_DBG("Partial data received (%zd bytes)", rsp->data_len);
+	} else if (final_data == HTTP_DATA_FINAL) {
+		NET_DBG("All the data received (%zd bytes)", rsp->data_len);
+	}
+
+	NET_INFO("Response status %s", rsp->http_status);
+
+	*status = rsp->http_status_code;
+}
+
+static int exec_http_query(struct net_if *iface, int sock)
+{
+	int32_t timeout = HTTP_WAIT_TIME;
+	struct http_request req;
+	uint16_t status;
+	int ret;
+
+	memset(&req, 0, sizeof(req));
+
+	req.method = HTTP_GET;
+	req.url = URL;
+	req.host = online_check->host;
+	req.port = online_check->port;
+	req.protocol = "HTTP/1.1";
+	req.response = response_cb;
+	req.recv_buf = online_check->http_recv_buf;
+	req.recv_buf_len = sizeof(online_check->http_recv_buf);
+
+	/* Make sure that we cannot access recv_buf data multiple times */
+	k_mutex_lock(&conn_mgr_mon_lock, K_FOREVER);
+
+	ret = http_client_req(sock, &req, timeout, &status);
+	if (ret < 0) {
+		NET_DBG("HTTP request failed (%d)", ret);
+		goto out;
+	}
+
+	/* Lets assume online also for 301 "Moved Permanently" */
+	if (status == 200 || status == 301) {
+		NET_DBG("Sending Online Connectivity event for interface %d",
+			net_if_get_by_iface(iface));
+		net_mgmt_event_notify(NET_EVENT_CONNECTIVITY_ONLINE, iface);
+	} else {
+		NET_DBG("Received HTTP status %d, not considering online.",
+			status);
+	}
+
+out:
+	k_mutex_unlock(&conn_mgr_mon_lock);
+
+	return ret;
+}
+
+static int do_online_http_check(struct net_if *iface, const char *host)
+{
+	int proto = IPPROTO_TCP;
+	const sec_tag_t *tags = NULL;
+	size_t tags_size = 0;
+	socklen_t addrlen;
+	int sock;
+	int ret;
+
+	if (!online_check->hostaddr_valid) {
+		ret = resolve_url(host);
+		if (ret < 0) {
+			NET_DBG("Cannot parse URL \"%s\" (%d)",
+				URL, ret);
+			return ret;
+		}
+	}
+
+	if (!online_check->hostaddr_valid) {
+		return -EAGAIN;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_SOCKOPT_TLS) && online_check->is_tls) {
+		if (online_check->cb == NULL) {
+			NET_DBG("HTTPS request but callback not registered. "
+				"HTTP online check disabled.");
+			return -ENOTSUP;
+		}
+
+		ret = online_check->cb(iface, &tags, &tags_size,
+				       host,
+				       online_check->host,
+				       online_check->port,
+				       &online_check->hostaddr,
+				       online_check->user_data);
+		if (ret < 0) {
+			NET_DBG("Setting up socket failed (%d)", ret);
+			return ret;
+		}
+
+		if (tags_size == 0 || tags == NULL) {
+			NET_DBG("Security tags is empty, TLS disabled");
+		} else {
+			proto = IPPROTO_TLS_1_2;
+		}
+	}
+
+	sock = zsock_socket(online_check->hostaddr.sa_family,
+			    SOCK_STREAM, proto);
+	if (sock < 0) {
+		NET_DBG("Socket creation failed (%d)", sock);
+		return ret;
+	}
+
+	if (IS_ENABLED(CONFIG_NET_SOCKETS_SOCKOPT_TLS) && proto == IPPROTO_TLS_1_2) {
```

Comment:
I would say, if we keep the HTTP connection check method, we do not want to connect. If we insist on establishing a full connection, then I have a hard time justifying this connection check method as compared to the end-user just attempting whatever connection type they ultimately intend to use with whatever credentials they ultimately intend to use

Reply:
IMHO it is very important that we make a successful connection, I do not see any reason why we would not do that.
User can use non-TLS connection if certs cannot be used.

Reply:
> I do not see any reason why we would not do that

Well, it could reduce the power used by the check, and it could also reduce the complexity of the required setup application-side. And it would also change my above argument about "why would the application want to do this?", since it might be a lower-cost pre-check than the actual connection, at least in theory.